// ng2-typewriter/src/typewriter.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, EventEmitter, Output } from "@angular/core";
var TypewriterComponent = (function () {
    function TypewriterComponent() {
        this.contents = [];
        this.erasable = false;
        this.beforeType = 500;
        this.beforeStart = 0;
        this.afterEnd = 1000;
        this.delay = 0;
        this.speed = 100 - Math.random() * 80;
        this.deletingAcceleration = this.speed / 8;
        this.deletingTopSpeed = this.speed / 4;
        this.cursor = "_";
        this.cursorDelay = -1;
        this.isDone = false;
        this.afterDone = new EventEmitter();
    }
    TypewriterComponent.prototype.ngOnInit = function () { };
    TypewriterComponent.prototype.onDone = function (isDone) {
        var _this = this;
        if (this.cursorDelay > 0) {
            setTimeout(function () {
                _this.isDone = isDone;
                _this.afterDone.emit(_this.isDone);
            }, this.cursorDelay);
        }
    };
    return TypewriterComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Array)
], TypewriterComponent.prototype, "contents", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], TypewriterComponent.prototype, "erasable", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "beforeType", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "beforeStart", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "afterEnd", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "delay", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "speed", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "deletingAcceleration", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "deletingTopSpeed", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], TypewriterComponent.prototype, "cursor", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], TypewriterComponent.prototype, "cursorDelay", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], TypewriterComponent.prototype, "afterDone", void 0);
TypewriterComponent = __decorate([
    Component({
        selector: "typewriter",
        template: "\n        <div g-typewriter [contents]=\"contents\" [speed]=\"speed\" [beforeType]=\"beforeType\" [beforeStart]=\"beforeStart\" [delay]=\"delay\"\n        [afterEnd]=\"afterEnd\" [erasable]=\"erasable\" [deletingAcceleration]=\"deletingAcceleration\" [deletingTopSpeed]=\"deletingTopSpeed\" (isDone)=\"onDone($event)\">\n            <span class=\"wrap\"></span>\n            <g-cursor [activated]=\"!isDone\">{{cursor}}</g-cursor>\n        </div>\n    ",
        styles: ["\n        .special-typewriter {\n            color       : darkred;\n            font-family : Arial, Helvetica, sans-serif;\n            font-weight : bolder;\n        }\n    "]
    }),
    __metadata("design:paramtypes", [])
], TypewriterComponent);
export { TypewriterComponent };
